---
title: "EssentialGeneVariability"
author: "Peter Shen"
date: '2017-11-21'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(dplyr)
```

## Data Import

### RNA Seq Data Processing
Import the sample data, which was produced taking 1000 lines from the RNASeq data file:
`head GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_tpm.gct 1000`
```{r}
gx <- data.table::fread("../GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_tpm.gct", sep = "\t", skip=2, nrows = 1000) # skip headers
```

Number of samples:
```{r}
ncol(gx) - 2 # subtract `name` and `description` fields
```

Structure of RNA Seq data (sample data):
* Name = The gene code in Emsembl (eg. http://www.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000223972;r=1:11869-14409)
* Description = The gene name
* Then the rest of the columns are tissue sample IDs `SAMPID` and their gene TPM reads

How the sample IDs are mapped to patients:
* The GTEX-#### is the Subject ID `SUBJID`
* The rest of the code is for the tissue type

### Phenotype Metadata
Import the phenotype data that contains the subject phenotypes:
```{r}
pheno <- read.csv("GTEx_v7_Annotations_SubjectPhenotypesDS.txt", sep="\t")
```
The `SUBJID` maps to the `GTEX-####` in the RNA Seq data file.

### Sample Attributes Metadata
Import the attributes data:
```{r}
attr <- read.csv("GTEx_v7_Annotations_SampleAttributesDS.txt", sep="\t")
```

Data dictionary of the attributes:

"SMATSSCR"="Autolysis" ("0"="None", "1"="Mild", "2"="Moderate", "3"="Severe")
"SAMPID"="Sample ID", "SMATSSCR"="Autolysis Score",
"SMNABTCH"="Nucleic Acid Isolation Batch ID",
"SMNABTCHT"="Type of nucleic acid isolation batch",
"SMNABTCHD"="Date of nucleic acid isolation batch",
"SMGEBTCH"="Genotype or Expression Batch ID",
"SMGEBTCHD"="Date of genotype or expression batch",
"SMGEBTCHT"="Type of genotype or expression batch",
"SMCENTER"="BSS collection site",
"SMPTHNTS"="Pathology Notes", "SMRIN"="RIN Number",
"SMTS"="Tissue Type (area of retrieval)",
"SMTSD"="Tissue Type (detail)", "SMUBRID"="Uberon ID",
"SMTSISCH"="Total Ischemic time",
"SMTSPAX"="PAXgene fixative time",
"SMTSTPTREF"="Period of sample procurement",
"SMAFRZE"="Samples in GTEx Analysis Freeze",
"SMGTC"="Genotype GTC file",
"SME2MPRT"="End 2 Mapping Rate",
"SMCHMPRS"="Chimeric Pairs",
"SMNTRART"="Intragenic Rate",
"SMNUMGPS"="Number of Gaps", "SMMAPRT"="Mapping Rate",
"SMEXNCRT"="Exonic Rate",
"SM550NRM"="5' 50-based normalization",
"SMGNSDTC"="Genes Detected",
"SMUNMPRT"="Unique Rate of Mapped",
"SM350NRM"="3' 50-base normalization",
"SMRDLGTH"="Read Length",
"SMMNCPB"="Mean Coverage Per Base",
"SME1MMRT"="End 1 Mismatch Rate",
"SMSFLGTH"="Fragment Length StdDev",
"SMESTLBS"="Estimated library size",
"SMMPPD"="Total mapped reads",
"SMNTERRT"="Intergenic Rate",
"SMRRNANM"="rRNA reads", "SMRDTTL"="Total reads",
"SMVQCFL"="Failed Vendor QC Check",
"SMMNCV"="Mean coefficient of variation",
"SMTRSCPT"="Transcripts Detected",
"SMMPPDPR"="Mapped Pairs",
"SMCGLGTH"="Cumulative Gap Length",
"SMGAPPCT"="Gap Percentage",
"SMUNPDRD"="Unpaired Reads", "SMNTRNRT"="Intronic Rate",
"SMMPUNRT"="Mapped Unique Rate of Total",
"SMEXPEFF"="Expression Profiling Efficiency",
"SMMPPDUN"="Mapped Unique",
"SME2MMRT"="End 2 Mismatch Rate",
"SME2ANTI"="End 2 Antisense",
"SMALTALG"="Alternative Aligments",
"SME2SNSE"="End 2 Sense",
"SMMFLGTH"="Fragment Length Mean",
"SMSPLTRD"="Split Reads", "SME1ANTI"="End 1 Antisense",
"SMBSMMRT"="Base Mismatch Rate",
"SME1SNSE"="End 1 Sense",
"SME1PCTS"="End 1 % Sense", "SMRRNART"="rRNA Rate",
"SME1MPRT"="End 1 Mapping Rate",
"SMNUM5CD"="Number Covered 5'",
"SMDPMPRT"="Duplication Rate of Mapped",
"SME2PCTS"="End 2 % Sense"

Source: https://github.com/nuno-agostinho/psichomics/blob/master/R/formats_gtexSampleInfo.R

## Preprocessing
```{r}
mat <- as.matrix(gx[, 3:ncol(gx)])
hist(mat)
```

We need to pull these data points closer together with log:
```{r}
gx[, 3:ncol(gx)] <- log2(gx[, 3:ncol(gx)]+1)

# look at the data again
mat <- as.matrix(gx[, 3:ncol(gx)])
hist(mat)
```
Now it seems to be more normal.

### Select the essential and non-essential genes from the dataset
gx_by_gene structure:
- SAMPID: sample ID
- SMTS: tissue type
- The rest cols are gene names with gene expressions
```{r}
gx_by_gene <- as.data.frame(t(gx[,3:ncol(gx)])) # select only expression data and transpose
colnames(gx_by_gene) <- gx$Description # gene names on columns
gx_by_gene$SAMPID<- names(gx)[3:ncol(gx)] # add another column for sample ID

# select sample ID and tissue type from attr and right join to gx_by_gene by sample ID
gx_by_gene <- attr %>%
    select(`SAMPID`,`SMTS`) %>%
        right_join(gx_by_gene)
```

Read in essential and non-essential gene lists:
```{r}
ess_gene_list <- data.table::fread("names_essential_genes.csv", drop = 1)$x
noness_gene_list <- data.table::fread("names_nonessential_genes.csv", drop = 1)$x
```

Get the essential and non-essential genes column IDs
```{r}
matched_ess <- match(ess_gene_list, colnames(gx_by_gene))
names(matched_ess) <- ess_gene_list
matched_ess <- matched_ess[!is.na(matched_ess)] # remove NA (no matching genes)

matched_non_ess <- match(noness_gene_list, colnames(gx_by_gene))
names(matched_non_ess) <- noness_gene_list
matched_non_ess <- matched_non_ess[!is.na(matched_non_ess)] # remove NA (no matching genes)
```

Get the expression data for the essential and non-essential genes
```{r}
ess_gx <- gx_by_gene[,c(1,2,matched_ess)]
# write.table(ess_gx, file="../essential_gene_data.csv", sep="\t")

non_ess_gx <- gx_by_gene[,c(1,2,matched_non_ess)]
# write.table(non_ess_gx, file="../non_essential_gene_data.csv", sep="\t")
```


Chen Dong
##ANOVA and POSTHOC analysis

In the following, I pick DDX11L1 as an example.
```{r}
# select one gene with the tissue types and its expressions
ess_gene_tissue_exp <- gx_by_gene[, c(2, 3)]
str(ess_gene_tissue_exp)
```

Hypothesis: The essential genes are equally expressed across all cell lines.
Alt hypothesis: Essential genes are differentially expressed across cell lines.
Run ANOVA and assess whether there are differences between tissue type and gene expression in the selected essential gene:
```{r}
ess_gene_anova <-
  lm(ess_gene_tissue_exp$DDX11L1 ~ ess_gene_tissue_exp$SMTS)
anova(ess_gene_anova)
```
Pr(>F) is much smaller than 0.05, meaning we can reject the null of no diff across cell lines.

Then we want to run posthoc to see where the differences lie between the cell lines.
```{r}
# Fit an analysis of variance model
a1 <- aov(ess_gene_tissue_exp$DDX11L1 ~ ess_gene_tissue_exp$SMTS)
# Compute Tukey Honest Significant Differences
posthoc <-
  TukeyHSD(x = a1,
           'ess_gene_tissue_exp$SMTS',
           conf.level = 0.95)
# Get the p_values between each pair of the cell lines
posthoc.p_values <- posthoc$`ess_gene_tissue_exp$SMTS`[, 4]
# Get those less than 0.05 but greater than 0
posthoc.p_values[which(posthoc.p_values < 0.05 &
                         posthoc.p_values > 0)]
```
The resulting pairs has p_value less than 0.05, meaning the differences between them are significant. 


## Jimmy - Run t-test for each gene

read the expression data
```{r}
ess_gx <- read.csv("expdata_essential_gene.csv",sep='\t')
non_ess_gx <- read.csv("expdata_nonessential_gene_data.csv", sep = '\t')
ess_gene_list <- data.table::fread("names_essential_genes.csv", drop = 1)$x
noness_gene_list <- data.table::fread("names_nonessential_genes.csv", drop = 1)$x
```

```{r}
tissue_list <- unique(ess_gx$SMTS)

# normalize method - make all values lie between 0 - 1
normalize <- function(exp){
  exp <- (exp - min(exp))/(max(exp) - min(exp))
  exp
}

# The function that perform t-test
get_pvalue_table <- function(type,filter.zero = FALSE, normalized = FALSE){
  
  type <- "non_essential"
  normalized <- TRUE
  
  gene_tissue_exp <- c()
  gene_list <- c()
  
  # Decide the data we use according to type argument
  if(type == "essential"){
    gene_tissue_exp <- ess_gx[,-1]
    gene_list <- ess_gene_list
  }else{
    gene_tissue_exp <- non_ess_gx[,-1]
    gene_list <- noness_gene_list
  }
  
  # Make all the value lie between 0 - 1 if normalized argument is true
  if(normalized == TRUE){
    for(tissue in tissue_list){
        for(gene in gene_list){
          gene_tissue_exp[which(gene_tissue_exp$SMTS==tissue),gene] <- normalize(gene_tissue_exp[which(gene_tissue_exp$SMTS==tissue),gene])
        }
    
    }
    gene_tissue_exp
  }
  
  gene_tissue_pvalue_table <- matrix(ncol = length(tissue_list),nrow = 0)
  # Get pvalue list of all the gene
  for(gene in gene_list){
    
  # for every gene, select one cell line as one group, and the remaining cell lines as another
    tissue_pvalue_list <- c()
    for(tissue in tissue_list){
      target_tissue_gene_exp <- gene_tissue_exp[which(gene_tissue_exp$SMTS==tissue),gene]
      
      remaining_tissue_gene_exp <- gene_tissue_exp[which(gene_tissue_exp$SMTS!=tissue),gene]
     
      # Filter zero value if filter.zero == TRUE
      if(filter.zero == TRUE){
        target_tissue_gene_exp <- target_tissue_gene_exp[which(target_tissue_gene_exp!=0)]
        remaining_tissue_gene_exp <- remaining_tissue_gene_exp[which(remaining_tissue_gene_exp!=0)]
      }
      
      # Perform t-test of 2 groups
      pval <- 0
      if(length(target_tissue_gene_exp) <= 1 | length(remaining_tissue_gene_exp) <= 1){
        pval <- NA
      }else{
        pval <- t.test(target_tissue_gene_exp, remaining_tissue_gene_exp)$p.value
      }
      tissue_pvalue_list <- append(tissue_pvalue_list,pval)
    }
    
    gene_tissue_pvalue_table <- rbind(gene_tissue_pvalue_table, tissue_pvalue_list)
  }
  
  # Get the final pvla
  gene_tissue_pvalue_table <- as.data.frame(gene_tissue_pvalue_table)
  colnames(gene_tissue_pvalue_table) <- tissue_list
  rownames(gene_tissue_pvalue_table) <- gene_list
  
  gene_tissue_pvalue_table

}
```

```{r}
    # Create subfolders first before creating the file
    dir.create((paste0("../ttest_results/")),showWarnings = FALSE)
    
    ess_ttest_results <- get_pvalue_table("essential")
    ess_ttest_results_filtered <- get_pvalue_table("essential", filter.zero = TRUE)
    ess_ttest_results_normalized <- get_pvalue_table("essential", normalized = TRUE)
    
    write.csv(ess_ttest_results, file="../ttest_results/essential_ttest_results.csv",row.names = TRUE)
    write.csv(ess_ttest_results_filtered,file="../ttest_results/ess_ttest_results_filtered.csv",row.names = TRUE)
    write.csv(ess_ttest_results_normalized, file="../ttest_results/essential_ttest_results_normalized.csv",row.names = TRUE)
```

Write all of the table
```{r}
    non_ess_ttest_results <- get_pvalue_table("non_essential")
    non_ess_ttest_results_filtered <- get_pvalue_table("non_essential", filter.zero = TRUE)
    non_ess_ttest_results_normalized <- get_pvalue_table("non_essential", normalized = TRUE)
    
    write.csv(non_ess_ttest_results, file="../ttest_results/non_essential_ttest_results.csv",row.names = TRUE)
    write.csv(non_ess_ttest_results_filtered,file="../ttest_results/non_ess_ttest_results_filtered.csv",row.names = TRUE)
    write.csv(non_ess_ttest_results_normalized, file="../ttest_results/non_essential_ttest_results_normalized.csv",row.names = TRUE)
```





